modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C

```diff
@@ -0,0 +1,1310 @@
+          res += _mgd.basis_species_charge[i] * _bulk_moles[i] / _mgd.basis_species_charge[basis_i];
+      }
+    }
+    else
+      res += -_bulk_moles[basis_i] + _basis_molality[0] * _basis_molality[basis_i];
+    for (unsigned eqm_j = 0; eqm_j < _num_eqm; ++eqm_j)
+      res += _basis_molality[0] * _mgd.eqm_stoichiometry(eqm_j, basis_i) * _eqm_molality[eqm_j];
+    return res;
+  }
+
+  // else: residual for surface potential
+  const unsigned sp = algebraic_ind - _num_basis_in_algebraic_system;
+  Real res = surfacePotPrefactor(sp) * (_surface_pot_expr[sp] - 1.0 / _surface_pot_expr[sp]);
+  for (unsigned j = 0; j < _num_eqm; ++j)
+    if (_mgd.surface_sorption_related[j] && _mgd.surface_sorption_number[j] == sp)
+      res += _basis_molality[0] * _mgd.eqm_species_charge[j] * _eqm_molality[j];
+  return res;
+}
+
+void
+EquilibriumGeochemicalSystem::computeJacobian(const DenseVector<Real> & res,
+                                              DenseMatrix<Real> & jac) const
+{
+  /*
+Note that in this function we make use of the fact that species can only be in the algebraic
+system if their molality is unknown (or in the case of water, the mass of solvent mass is
+unknown).  This means that the molality of the equilibrium species depends on
+(activity_coefficient * basis molality)^stoi, so that the derivative is quite simple.  Eg, we
+don't have to worry about derivatives with respect to fixed-activity things, or fixed fugacity.
+Also, note that the constructor and the various "set" methods of this class enforce molality > 0,
+so there are no division-by-zero problems.  Also, note that we never compute derivatives of the
+activity coefficients, or the activity of water with respect to the molalities, as they are
+assumed to be quite small.  Finally, the surface_pot_expr will also always be positive.
+  */
+  // correctly size and zero
+  jac.resize(_num_in_algebraic_system, _num_in_algebraic_system);
+  const Real nw = _basis_molality[0];
+
+  // jac(a, b) = d(R_a) / d(m_b), where a corresponds to a molality
+  for (unsigned a = 0; a < _num_basis_in_algebraic_system; ++a)
+  {
+    const unsigned basis_of_a = _basis_index[a];
+    for (unsigned b = 0; b < _num_basis_in_algebraic_system; ++b)
+    {
+      const unsigned basis_of_b = _basis_index[b];
+      if (basis_of_b == 0) // special cases: mass of solvent water is an unknown
+      {
+        if (basis_of_a != _charge_balance_basis_index)
+          jac(a, 0) += (res(a) + _bulk_moles[basis_of_a]) / nw;
+        else
+        {
+          Real extra_term = 0.0;
+          for (unsigned i = 0; i < _num_basis; ++i)
+          {
+            if (i == _charge_balance_basis_index)
+              continue;
+            else if (_mgd.basis_species_charge[i] ==
+                     0.0) // certainly includes water, minerals and gases
+              continue;
+            else if (_in_algebraic_system[i]) // bulk moles must have been provided by user
```

Comment:
else instead of else if?

Reply:
I *think* this should be `else if`.   I'm happy to be wrong!  Did you work out the jacobian and think it should be `else` ?

Reply:
I couldn't tell - you are probably correct!